rpctest/memwallet.go

```
@@ -0,0 +1,517 @@
+func (m *memWallet) start() {
+	go m.chainSyncer()
+}
+
+// syncedHeight returns the height the wallet is known to be synced to.
+//
+// This function is safe for concurrent access.
+func (m *memWallet) syncedHeight() int32 {
+	m.RLock()
+	defer m.RUnlock()
+	return m.currentHeight
+}
+
+// setRPCClient saves the passed rpc connection to btcd as the wallet's
+// personal rpc connection.
+func (m *memWallet) setRPCClient(rpcClient *btcrpcclient.Client) {
+	m.rpc = rpcClient
+}
+
+// ingestBlock is a call-back which is to be triggered each time a new block is
+// connected to the main chain. Ingesting a block updates the wallet's internal
+// utxo state based on the outputs created and destroyed within each block.
+func (m *memWallet) ingestBlock(blockHash *chainhash.Hash, height int32, t time.Time) {
+	// Append this new chain update to the end of the queue of new chain
+	// updates.
+	m.chainMtx.Lock()
+	m.chainUpdates = append(m.chainUpdates, &chainUpdate{blockHash, height})
+	m.chainMtx.Unlock()
+
+	// Launch a goroutine to signal the chainSyncer that a new update is
+	// available. We do this in a new goroutine in order to avoid blocking
+	// the main loop of the rpc client.
+	go func() {
+		m.chainUpdateSignal <- struct{}{}
+	}()
+}
+
+// chainSyncer is a goroutine dedicated to processing new blocks in order to
+// keep the wallet's utxo state up to date.
+//
+// NOTE: This MUST be run as a goroutine.
+func (m *memWallet) chainSyncer() {
+	var update *chainUpdate
+
+	for range m.chainUpdateSignal {
+		m.chainMtx.Lock()
+
+		// A new update is available, so pop the new chain update from
+		// the front of the update queue.
+		update = m.chainUpdates[0]
+		m.chainUpdates[0] = nil // Set to nil to prevent GC leak.
+		m.chainUpdates = m.chainUpdates[1:]
+
+		m.chainMtx.Unlock()
+
+		// Fetch the new block so we can process it shortly below.
+		block, err := m.rpc.GetBlock(update.blockHash)
+		if err != nil {
+			return
```

Hmm yeah initially I considered passing in an instance of `testing.T` would which fail the test if an error was detected. But then I realized the package may not always be used in conjunction with a `testing.T`. 

I ended up just silently failing here, since erroneous behavior should lead to a failure in the test suite as you pointed out.